File: leetcode/LinkedList/leetcode-0002/2.go
package linkedlist

// ListNode Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := new(ListNode)
	pre := dummy
	sum, tmp := 0, 0

	for l1 != nil && l2 != nil {
		sum = l1.Val + l2.Val + tmp
		tmp = sum / 10
		sum %= 10
		node := &ListNode{sum, nil}
		pre.Next = node
		pre = pre.Next
		l1 = l1.Next
		l2 = l2.Next
	}

	for l1 != nil {
		sum = l1.Val + tmp
		tmp = sum / 10
		sum %= 10
		node := &ListNode{sum, nil}
		pre.Next = node
		pre = pre.Next
		l1 = l1.Next
	}

	for l2 != nil {
		sum = l2.Val + tmp
		tmp = sum / 10
		sum %= 10
		node := &ListNode{sum, nil}
		pre.Next = node
		pre = pre.Next
		l2 = l2.Next
	}

	if tmp != 0 {
		node := &ListNode{tmp, nil}
		pre.Next = node
	}
	return dummy.Next
}

File: leetcode/LinkedList/leetcode-0002/2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode pre = dummy;

        int tmp = 0;
        while (l1 != null && l2 != null) {
            int sum = l1.val + l2.val + tmp;
            tmp = sum / 10;
            sum %= 10;
            ListNode node = new ListNode(sum);
            pre.next = node;
            pre = pre.next;
            l1 = l1.next;
            l2 = l2.next;
        }

        while (l1 != null) {
            int sum = l1.val + tmp;
            tmp = sum / 10;
            sum %= 10;
            ListNode node = new ListNode(sum);
            pre.next = node;
            pre = pre.next;
            l1 = l1.next;
        }

        while (l2 != null) {
            int sum = l2.val + tmp;
            tmp = sum / 10;
            sum %= 10;
            ListNode node = new ListNode(sum);
            pre.next = node;
            pre = pre.next;
            l2 = l2.next;
        }

        if (tmp != 0) {
            ListNode node = new ListNode(tmp);
            pre.next = node;
        }

        return dummy.next;
    }
}